core/src/main/java/org/primefaces/extensions/component/inputpin/InputPinRenderer.java

```diff
@@ -0,0 +1,169 @@
+        if (valueToRender == null) {
+            valueToRender = Constants.EMPTY_STRING;
+        }
+
+        encodeMarkup(context, inputPin, valueToRender);
+        encodeScript(context, inputPin);
+    }
+
+    protected void encodeMarkup(final FacesContext context, final InputPin inputPin, final String valueToRender)
+                throws IOException {
+        final ResponseWriter writer = context.getResponseWriter();
+        final String clientId = inputPin.getClientId(context);
+        final String styleClass = getStyleClassBuilder(context)
+                    .add(InputPin.STYLE_CLASS, inputPin.getStyleClass())
+                    .build();
+
+        writer.startElement("span", inputPin);
+        writer.writeAttribute("id", clientId, null);
+        writer.writeAttribute(Attrs.CLASS, styleClass, "styleClass");
+
+        if (inputPin.getStyle() != null) {
+            writer.writeAttribute(Attrs.STYLE, inputPin.getStyle(), Attrs.STYLE);
+        }
+
+        encodeInput(context, inputPin, clientId, valueToRender);
+        encodeHiddenInput(context, inputPin, clientId, valueToRender);
+
+        writer.endElement("span");
+    }
+
+    protected void encodeInput(final FacesContext context, final InputPin inputPin, final String clientId,
+                final String valueToRender)
+                throws IOException {
+
+        final ResponseWriter writer = context.getResponseWriter();
+        final String inputStyle = inputPin.getInputStyle();
+        final String inputStyleClass = createStyleClass(inputPin, "inputStyleClass", InputPin.CELL_STYLE_CLASS);
+        final char[] chars = valueToRender.toCharArray();
+        for (int i = 1; i <= inputPin.getSize(); i++) {
+
+            if (i > 1 && !LangUtils.isBlank(inputPin.getSeparator())) {
+                writer.writeText(inputPin.getSeparator(), InputPin.PropertyKeys.separator.name());
+            }
+
+            final String inputId = clientId + InputPin.INPUT_SUFFIX + i;
+            final String inputValue = chars.length >= i ? String.valueOf(chars[i - 1]) : "";
+
+            writer.startElement("input", null);
+            writer.writeAttribute("id", inputId, null);
+            writer.writeAttribute("name", inputId, null);
+            writer.writeAttribute("value", inputValue, null);
+            writer.writeAttribute("size", 1, null);
+            writer.writeAttribute(Attrs.CLASS, inputStyleClass, null);
+
+            if (!isValueBlank(inputStyle)) {
+                writer.writeAttribute(Attrs.STYLE, inputStyle, null);
+            }
+
+            if ("password".equalsIgnoreCase(inputPin.getType())) {
+                writer.writeAttribute("type", inputPin.getType(), null);
+            }
+
+            if (inputPin.isNumeric()) {
+                writer.writeAttribute("inputmode", "numeric", null);
+            }
+
+            writer.writeAttribute(HTML.ARIA_LABEL, MessageFactory.getMessageString(context, MESSAGE_ARIA_LABEL_KEY, i), null);
```

Comment:
I know this is easier here but we have been trying to move all ARIA Labels to the client side in JS code because it is easier to update for user and cleaner.  The mix of server side and client side labels was always a bad idea and in 14 we have moved "most" of the labels to the client side.

Reply:
ok, I will change

Reply:
Is there a localization file for js in the extensions project?